In [5]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine
import seaborn as sns
import matplotlib.pyplot as plt
from scipy.stats.stats import ttest_ind
import warnings

warnings.filterwarnings('ignore')

In [2]:
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'studentsperformance'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

df = pd.read_sql_query('select * from studentsperformance',con=engine)

# no need for an open connection, 
# as we're only doing a single query
engine.dispose()

In [3]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

# 1. Are there any differences between the genders, ethnicities, and parental level of education with respect to their performances in exams?

In [4]:
df.groupby('gender').mean()

math score  reading score  writing score
gender                                          
female   63.633205      72.608108      72.467181
male     68.728216      65.473029      63.311203

In [7]:
ttest_ind(df[df['gender'] == 'female'][['math score', 'reading score', 'writing score']], df[df['gender'] == 'male'][['math score', 'reading score', 'writing score']])

Ttest_indResult(statistic=array([-5.38324587,  7.95930801,  9.97955791]), pvalue=array([9.12018555e-08, 4.68053874e-15, 2.01987771e-22]))

According to the t-test there is siginificant differences between genders. Males seem to have higher math scores. Females seem to have higher reading and writing scores.

In [20]:
df.groupby('race/ethnicity').mean()

math score  reading score  writing score
race/ethnicity                                          
group A          61.629213      64.674157      62.674157
group B          63.452632      67.352632      65.600000
group C          64.463950      69.103448      67.827586
group D          67.362595      70.030534      70.145038
group E          73.821429      73.028571      71.407143

In [19]:
groups = df.groupby('race/ethnicity').mean().index
for i in range(len(groups)):
    for j in range(i+1, len(groups)):
        print('t-test between group {} and {}: \n{}\n'.format(groups[i], groups[j], ttest_ind(df[df['race/ethnicity'] == groups[i]][['math score', 'reading score', 'writing score']], df[df['race/ethnicity'] == groups[j]][['math score', 'reading score', 'writing score']])))

t-test between group group A and group B: 
Ttest_indResult(statistic=array([-0.93550763, -1.36336823, -1.46244206]), pvalue=array([0.3503416 , 0.17387355, 0.14475374]))

t-test between group group A and group C: 
Ttest_indResult(statistic=array([-1.59972203, -2.57543222, -2.8488724 ]), pvalue=array([0.11043811, 0.0103644 , 0.00461069]))

t-test between group group A and group D: 
Ttest_indResult(statistic=array([-3.34669824, -3.04685486, -4.15563164]), pvalue=array([9.06863605e-04, 2.48865446e-03, 4.08480573e-05]))

t-test between group group A and group E: 
Ttest_indResult(statistic=array([-5.93614103, -4.07110623, -4.22347366]), pvalue=array([1.08355912e-08, 6.46339593e-05, 3.48327806e-05]))

t-test between group group B and group C: 
Ttest_indResult(statistic=array([-0.73156699, -1.32231562, -1.59649634]), pvalue=array([0.46477089, 0.18665928, 0.1110011 ]))

t-test between group group B and group D: 
Ttest_indResult(statistic=array([-2.8284554 , -1.94515636, -3.19929333]), pvalue=ar

According to the t-test, there seems to be siginificant difference between some groups.

1. Between group A and group B, there is no significant difference in test scores.
2. Between group A and group C, group A seems to have lower test scores for reading and writing. There is no difference in math scores.
3. Between group A and group D, group A seems to have lower test scores in all subjects.
4. Between group A and group E, group A seems to have lower test scores in all subjects.
5. Between group B and group C, there is no significant difference in test scores.
6. Between group B and group D, group B seems to have lower test scores in math and writing. There is no significant difference in reading scores.
7. Between group B and group E, group B seems to have lower test scores in all subjects.
8. Between group C and group D, group C seems to have lower math scores. There is no significant difference in reading and writing.
9. Between group C and group E, group C seems to have lower test scores in all subjects.
10. Between group D and group E, group D seems to have lower test scores in math and reading. There is no significant difference in writing scores.

In [21]:
df.groupby('parental level of education').mean()

math score  reading score  writing score
parental level of education                                          
associate's degree            67.882883      70.927928      69.896396
bachelor's degree             69.389831      73.000000      73.381356
high school                   62.137755      64.704082      62.448980
master's degree               69.745763      75.372881      75.677966
some college                  67.128319      69.460177      68.840708
some high school              63.497207      66.938547      64.888268

In [23]:
groups = df.groupby('parental level of education').mean().index
for i in range(len(groups)):
    for j in range(i+1, len(groups)):
        print('t-test between group {} and {}: \n{}\n'.format(groups[i], groups[j], ttest_ind(df[df['parental level of education'] == groups[i]][['math score', 'reading score', 'writing score']], df[df['parental level of education'] == groups[j]][['math score', 'reading score', 'writing score']])))

t-test between group associate's degree and bachelor's degree: 
Ttest_indResult(statistic=array([-0.87866419, -1.29779699, -2.11592966]), pvalue=array([0.38020758, 0.19524206, 0.03508162]))

t-test between group associate's degree and high school: 
Ttest_indResult(statistic=array([3.94813253, 4.53802142, 5.34872564]), pvalue=array([9.24731427e-05, 7.44216780e-06, 1.46505455e-07]))

t-test between group associate's degree and master's degree: 
Ttest_indResult(statistic=array([-0.84112297, -2.19120158, -2.78124155]), pvalue=array([0.40099936, 0.02926415, 0.00578403]))

t-test between group associate's degree and some college: 
Ttest_indResult(statistic=array([0.54268427, 1.11231882, 0.76159686]), pvalue=array([0.58761828, 0.26660029, 0.44670317]))

t-test between group associate's degree and some high school: 
Ttest_indResult(statistic=array([2.82005398, 2.71836022, 3.33172235]), pvalue=array([0.00504108, 0.00684677, 0.0009436 ]))

t-test between group bachelor's degree and high school: 

There are some significant between some groups

# 2. Are there any differences between the lunch types with respect to their performances in exams? If there are, how do you explain this?

In [25]:
df.groupby('lunch').mean()

math score  reading score  writing score
lunch                                                 
free/reduced   58.921127      64.653521      63.022535
standard       70.034109      71.654264      70.823256

In [26]:
ttest_ind(df[df['lunch'] == 'free/reduced'][['math score', 'reading score', 'writing score']], df[df['lunch'] == 'standard'][['math score', 'reading score', 'writing score']])

Ttest_indResult(statistic=array([-11.83718047,  -7.45105647,  -8.0097842 ]), pvalue=array([2.41319560e-30, 2.00279665e-13, 3.18618958e-15]))

There appears to be a significant difference in scores between lunch groups. Students with standard lunch seem to have higher test scores than students with free/reduced lunch. This may be due to the socioecominic advantages of the students who can afford standard lunch.

# 3. Does the test preparation course seem to have an effect on the exam performances?

In [27]:
df.groupby('test preparation course').mean()

math score  reading score  writing score
test preparation course                                          
completed                 69.695531      73.893855      74.418994
none                      64.077882      66.534268      64.504673

In [28]:
ttest_ind(df[df['test preparation course'] == 'completed'][['math score', 'reading score', 'writing score']], df[df['test preparation course'] == 'none'][['math score', 'reading score', 'writing score']])

Ttest_indResult(statistic=array([ 5.70461642,  7.87166354, 10.40917344]), pvalue=array([1.53591346e-08, 9.08178334e-15, 3.68529174e-24]))

The test preparation course does seem to make a significant difference in exam scores. Students who completed the test preparation course have significantly higher exam scores than students who did not complete the course.

# 4. Which two exam scores are correlated the most with each other?

In [29]:
df.corr()

math score  reading score  writing score
math score       1.000000       0.817580       0.802642
reading score    0.817580       1.000000       0.954598
writing score    0.802642       0.954598       1.000000

READING SCORES and WRITING SCORES have the highest correlation coefficient of 0.954598.